### ISSUE TITLE
[Image] Old images on Movies example app listings

### ISSUE THREAD

**Issue by jchris** [facebook/react-native#89]
To reproduce, search for a movie and browse into it, then back out of the movie view to the search, then delete your search terms by holding down the delete key (on your laptop while running in simulator). It seems to be timing dependent but with a few tries I was able to generate the screenshots below.

![wayne](https://cloud.githubusercontent.com/assets/253/6314970/4b5075e8-b9a7-11e4-9410-e9f018ea076c.png)

![apollo](https://cloud.githubusercontent.com/assets/253/6314971/4ff834dc-b9a7-11e4-9987-259565eb8111.png)

Maybe this is throttling from the data API? Maybe caching would help?

![not found](https://cloud.githubusercontent.com/assets/253/6314977/78c6f812-b9a7-11e4-813d-a1f5782a601f.png)

**Comment by spicyj** [comment]
The "keep a" experience is probably better at an infrastructural level because you can always get the "blank" way by adding a key to the Image equal to the src.

**Comment by plredmond** [comment]
@vjeux I've tried using `<Image defaultSource=... />` to show some kind of a loading placeholder image, but it shows the old image indefinitely. I'll get a minimal repro set up for you.

Meanwhile is there workaround we can use to blank the image and force loading the correct one?

**Comment by plredmond** [comment]
workaround

```es6
/* @flow */
'use strict';

var React = require('react-native');
var { Image } = React;

// Hack to work around https://github.com/facebook/react-native/issues/89
//
// Render an <Image /> normally.
// When react reuses the component with a new url, force react to render a
// blank image once before rendering the correct url later.

var DelayedImage = React.createClass({
  propTypes: Image.propTypes,
  getInitialState(): { showImage: boolean } {
    return { showImage: true };
  },
  componentWillReceiveProps: function(nextProps: any) {
    if(this.props.source.uri != nextProps.source.uri) {
      this.setState({ showImage: false });
      setTimeout(() => this.setState({ showImage: true }), 0);
    }
  },
  render: function(): React.Component {
    return <Image {...this.props} source={{uri: this.state.showImage ? this.props.source.uri : null}} />
  },
});

module.exports = DelayedImage;
```

**Comment by brentvatne** [comment]
The issue here now is that the images change but the text does not

![bug](https://cloud.githubusercontent.com/assets/90494/7899847/683949c4-06ea-11e5-9111-9cf46a2b0a00.gif)

@vjeux - who is the owner of ListView?

**Comment by spicyj** [comment]
Probably an Image issue; I can't imagine it being specific to ListView.

**Comment by sahrens** [comment]
This is pretty busted - any updates, @nicklockwood?

**Comment by christopherdro** [comment]
This is a tough one to catch since the results are never consistent. I was still able to reproduce the issue, maybe 1 out of every 4-5 attempts, following @jchris explanation. See demo below...

I think the catch here is to make sure each `<Image />` has a key property set. 
In the case of the Movie Example, add a key to each image inside of our MovieCell seems to resolve the issue.

I've noticed some other people have issues with Images rendering properly within ListView and this did the trick.
```
<Image
  source={getImageSource(this.props.movie, 'det')}
  style={styles.cellImage}
  key={this.props.movie.id}
/>
```
Not sure if having to add a key  a bug or just a requirement to get the images to re-render properly.
Thoughts?

![Image of Demo]
(http://d.pr/i/16bcY)

**Comment by jchris** [comment]
My thought is that a "loadingImage" property? would force the state machine to represent the that the image isn't ready.

**Comment by brentvatne** [comment]
@jchris - we can do this now by leveraging the [onLoad](https://facebook.github.io/react-native/docs/image.html#onloadstart) callbacks :smile:

**Comment by MossP** [comment]
Quick observation on this. I've been seeing this bug for sometime now elsewhere. I've tried keying the image, setting it in state, then calling form there. I've dropped debug points and can see the url that was being assigned to the image wasn't what I was seeing. It only happened on one image in my render row though which got me thinking...

Through elimination the thing that got it working for me was to remove the `<TouchableOpacity/>` and replace with `<TouchableWithoutFeedback />`. I have no idea why this works and it may be a Red Herring (maybe something was cached?) but mine has worked consistently through 15 builds now, and I can trigger the bug again as soon as I swap back.
